lib/room_booker.rb

```diff
@@ -0,0 +1,172 @@
+
+      reservations.each do |reservation|
+        unless DateLogic.date_ranges_exclusive?(
+          reservation.check_in,
+          reservation.check_out,
+          check_in,
+          check_out
+        )
+
+          reserved_rooms << reservation.room
+        end
+      end
+
+      unavailable_rooms = reserved_rooms + find_blocked_rooms(check_in, check_out)
+
+      available_rooms = rooms.dup
+
+      unavailable_rooms.each do |unavailable_room|
+        available_rooms.delete_if { |room| room.id == unavailable_room.id }
+      end
+
+      return available_rooms
+    end
+
+    def new_reservation(room_id, check_in, check_out)
+      room = find_room_by_id(room_id)
+      room_unavailable?(room, check_in, check_out)
+      new_reservation = BookingLogic::Reservation.new(room, check_in, check_out)
+      reservations << new_reservation
+      return new_reservation
+    end
+
+    def room_unavailable?(room, check_in, check_out)
+      unless list_available_rooms(check_in, check_out).include?(room)
+        raise RoomNotAvailable, "Room not available for the given dates"
+      end
+    end
+
+    def find_block(name)
+      found_block = blocks.find { |block| block.name == name }
+      return found_block
+    end
+
+    def find_blocked_rooms(check_in, check_out)
+      blocked_rooms = []
+
+      blocks.each do |block|
+        unless DateLogic::date_ranges_exclusive?(
+          block.check_in,
+          block.check_out,
+          check_in,
+          check_out
+        )
+
+          block.rooms.each do |room|
+            blocked_rooms << room
+          end
+        end
+      end
+
+      return blocked_rooms
+    end
+
+    def block_available_rooms(check_in, check_out, number_of_rooms, rate)
+      
+      available_rooms = list_available_rooms(check_in, check_out)
+
+      room_block = []
+
+      number_of_rooms.times do |i|
+        room_block << available_rooms[i].dup
+      end
+
+      if room_block.length < number_of_rooms
+        raise StandardError, "There are not enough available rooms for the given dates to create this room block"
+      end
+
+      return room_block
+    end
+
+    def new_room_block(name, check_in, check_out, number_of_rooms, rate)
+
+      block_of_rooms = block_available_rooms(
+        check_in,
+        check_out,
+        number_of_rooms,
+        rate
+      )
+
+      new_room_block = BookingLogic::RoomBlock.new(
+        name,
+        check_in,
+        check_out,
+        block_of_rooms,
+        rate
+      )
+
+      new_room_block.set_blocked_room_rate
+
+      @blocks << new_room_block
+      return new_room_block
+    end
+
+    def new_block_reservation(name)
+      block = find_block(name)
+      room = block.available.first
+      room.block_reserved = true
+      new_reservation = BookingLogic::Reservation.new(room, block.check_in, block.check_out)
+      reservations << new_reservation
+      return new_reservation
+    end
+
+    def set_room_rate(room_id, custom_rate)
```

Comment:
I think this method is unused? I think you probably wrote this and then realized you should move it to the RoomBlock class, which you did... but you left this code here! You can probably get rid of this if that's the case